File: captainhcg/go/198-house-robber.go
func rob(nums []int) int {
    if len(nums) == 0{
        return 0
    }
    if len(nums) == 1{
        return nums[0]
    }
    max_v := make([]int, len(nums))
    for idx := 0; idx < len(nums); idx += 1{
        if idx == 0{
            max_v[idx] = nums[0]
        } else if idx == 1{
            max_v[idx] = max(nums[0], nums[1])
        } else {
            max_v[idx] = max(max_v[idx-1], max_v[idx-2] + nums[idx])
        }
    }
    return max(max_v[len(nums)-1], max_v[len(nums)-2])
}

func max(a, b int) int {
    if a > b{
        return a
    }
    return b
}

File: bwu/dp/198-house-robber.py
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        if n == 0:
            return 0
        if n <= 2:
            return max(nums)
            
        dp = [[0, 0] for _ in xrange(n)]
        dp[0][0] = nums[0]
        
        for i in xrange(1, n):
            dp[i][0] = dp[i - 1][1] + nums[i]
            dp[i][1] = max(dp[i - 1][0], dp[i - 1][1])
            
        return max(dp[n - 1][0], dp[n - 1][1])
